FUSETOOLS2-725 - Removing single window limit on didact

bfitzpat opened the pull request:
Removing single-window limitation on opening Didact files. https://issues.redhat.com/browse/FUSETOOLS2-725

![multi-didact](https://user-images.githubusercontent.com/530878/103691597-a4299300-4f53-11eb-8bb7-890e4116fd00.jpg)

Exploring issues with caching on a per-window basis (see https://issues.redhat.com/browse/FUSETOOLS2-930)

Signed-off-by: Brian Fitzpatrick <bfitzpat@redhat.com>

bfitzpat commented:
Essentially now when the workspace shuts down we cache the details from the first Didact window closed in a given session. In this workspace's case, it opened with the Didact demo, so that's tucked away in the cache. So regardless of what you have open when the workspace closes, you get...

![multi-didact-cache-problem](https://user-images.githubusercontent.com/530878/103691789-04b8d000-4f54-11eb-89f5-3aa0138c6718.jpg)

I think maybe we can use the vscode setState/getState API to handle things differently on a per-window basis rather than having a single cache for all windows: https://code.visualstudio.com/api/extension-guides/webview#getstate-and-setstate

bfitzpat commented:
Definitely a work in progress, but wanted to stash the work.

bfitzpat commented:
Improvement... now have it working and saving html state on a per-window basis. 

* Removed all the history functionality, which is no longer necessary. 
* Updated tests
* Working on a small issue with the didact window title being reset for the first window -- only affects the first window though

![multi-didact-demo](https://user-images.githubusercontent.com/530878/104065148-7eeb8d80-51bc-11eb-9d77-a3143ebc41f1.gif)

apupier commented on src/didactWebView.ts:
```diff
@@ -405,33 +411,11 @@ export class DidactWebviewPanel {
 		}
 		if (this.currentHtml) {
-			DidactWebviewPanel.cacheFile(); // update the cache with the new content
-			this._panel.webview.html = this.currentHtml;
+			if (this._panel && this._panel.webview && this._panel.active) {
```
the two if condition can be merged

bfitzpat commented:
Success!! 
![multi-didact-demo-26-JAN-2021](https://user-images.githubusercontent.com/530878/105901229-4c25ff80-5fda-11eb-990d-13e2556f61f3.gif)

bfitzpat commented:
This is a large commit, but the core of the work is in DidactManager and DidactPanel.

apupier reviewed (changes requested):
* I have not found tests for the new functionality
* every time I open a new VS Code isntance, another `Didact tutorial` page is opened
* The welcome page stays blank (other tutorial are loading)
* tried the "Writing your first Didact tutorial":
  * The "Click here to open the new myfirst.didact.md file" is no more working. The erro rmessage seems to say that the dots has been removed: `Unable to open 'didactmdfile.json': Unable to read file '/home/apupier/ws/test-didact/didactmdfile.json' (Error: Unable to resolve non-existing file '/home/apupier/ws/test-didact/didactmdfile.json').`
  * it simplifies a lot the workflow!
* When hitting Ctrl+Shift+V to preview a tutorial that i'm currently writing, it is opening a new tab each time. I woudl expect that for the same file, it is only refreshing the existing one (but sounds possible to have it as an improvement for another iteration)

bfitzpat commented on src/didactPanelSerializer.ts:
```diff
@@ -0,0 +1,40 @@
+		this._context = context;
+	}
+
+	public async deserializeWebviewPanel(webviewPanel: WebviewPanel, state: any) : Promise<void> {
```
Not sure it's needed. There are TWO items in this state, so it's very very minimal. If it was more complex, I might agree.

apupier replied:
deactivate the eslint rule which is forbiding usage of any then

bfitzpat replied:
@apupier I thought you were against deactivating eslint rules?

bfitzpat replied:
Added the necessary rule disablement

apupier replied:
I prefer not deactivating them without good reasons but the rules are not followed, it is then better to deactivate them. it avoids to have noise of rule that you choose to not follow.
It is better to introduce a type instead of disabling the rule. But you do not want. So it remains better to to disable the rule if it is not followed.
By not creatign the type, the type-safe part of TypeScript language is not leveraged. We will hit all the troubles of non-typed languages.

bfitzpat commented on src/didactPanel.ts:
```diff
@@ -0,0 +1,412 @@
+				// Enable javascript in the webview
+				enableScripts: true,
+
+				// And restrict the webview to only loading content from our extension's `media` directory.
+				localResourceRoots: localResourceRoots, 
```
The comment is misleading. Will modify or remove since we do handle adding a base directory if it's at a different location.

bfitzpat commented on src/didactManager.ts:
```diff
@@ -0,0 +1,76 @@
+import { DidactPanel } from './didactPanel';
+
+export const DEFAULT_TITLE_VALUE = `Didact Tutorial`;
+export const VIEW_TYPE = 'didact2';
```
Good catch. Leftover from when I had both didact web views in place.

bfitzpat commented on media/main.js:
```diff
@@ -70,17 +82,23 @@ function () {
 
+	function updateState() {
+		const textToCache = '<!DOCTYPE HTML>' + '\n' + document.documentElement.outerHTML;
+		const step1 = encodeURI(textToCache);
```
sure

bfitzpat replied:
Resolved, though not sure why this particular comment doesn't show as outdated.

bfitzpat commented:
> * I have not found tests for the new functionality

There really isn't new functionality. All we did was remove the limit on a single window. Not quite sure what kind of tests you expect for that.

> * every time I open a new VS Code isntance, another `Didact tutorial` page is opened

Can you clarify what's going on here? In my local testing, it doesn't open additional pages. Is this with the default setting in place to open the default tutorial set in settings?
 
> * The welcome page stays blank (other tutorial are loading)

Can you clarify what this means? Maybe provide a video/gif? 
 
>   * The "Click here to open the new myfirst.didact.md file" is no more working. The erro rmessage seems to say that the dots has been removed: `Unable to open 'didactmdfile.json': Unable to read file '/home/apupier/ws/test-didact/didactmdfile.json' (Error: Unable to resolve non-existing file '/home/apupier/ws/test-didact/didactmdfile.json').`

Will investigate.

>   * it simplifies a lot the workflow!

Great!
 
> * When hitting Ctrl+Shift+V to preview a tutorial that i'm currently writing, it is opening a new tab each time. I woudl expect that for the same file, it is only refreshing the existing one (but sounds possible to have it as an improvement for another iteration)

Interesting quirk with this implementation. I will see if there's a way to work around that, perhaps by storing the uri and comparing that when opening a new window so we only open the same one. But let's spin that off into a separate issue I think. Related, but separate.

apupier commented:
> > * I have not found tests for the new functionality
> 
> There really isn't new functionality. All we did was remove the limit on a single window. Not quite sure what kind of tests you expect for that.

that's a big functionality to be able to open 2 didact tutorial at the same time. Also that they keep the state. A test that is openign 2 tutorials and checkign that they both open with correct state woudl very useful.

apupier commented:
> > * The welcome page stays blank (other tutorial are loading)
> 
> Can you clarify what this means? Maybe provide a video/gif?

- Open VS code with Didact installed
- wait that the welcome page opens
- the page is blank

![Screenshot from 2021-01-27 15-24-24](https://user-images.githubusercontent.com/1105127/106004711-c6768280-60b3-11eb-899f-343882f31a28.png)

bfitzpat commented:
> that's a big functionality to be able to open 2 didact tutorial at the same time. Also that they keep the state. A test that is openign 2 tutorials and checkign that they both open with correct state woudl very useful.

I don't disagree with you. However, unless you know how to shut down a vscode instance during a test and re-open it, I'm afraid we're kind of out of luck there.

apupier commented:
> > that's a big functionality to be able to open 2 didact tutorial at the same time. Also that they keep the state. A test that is openign 2 tutorials and checkign that they both open with correct state woudl very useful.
> 
> I don't disagree with you. However, unless you know how to shut down a vscode instance during a test and re-open it, I'm afraid we're kind of out of luck there.

we can have at least teh check that 2 tutorials can be opened side by side.

Doe it really need a shutdown? Can it be a "reload window"? (for whcih VS Code has a command is used when some installation are installed)

bfitzpat commented:
I believe this latest commit addresses all of the suggestions and issues raised except for the one where if you open a single didact more than once, it opens more than one window so you can have duplicates. Working on that next.

bfitzpat commented:
Ok, with the last two commits we have a couple of things:
* now using the uri effectively to ensure that we don't open more than one window for a given URI 
* now testing to ensure that the we avoid duplicates 
* renaming the old webview tests to panel tests to better reflect the refactored work 

My question though is this: If I have a Didact tutorial open and re-open it, it overwrites any of the html changes that may have happened since the initial open. For instance, if I checked requirements, those requirements are now green or red with indications of whether the requirements are met. Those html changes are persisted behind the scenes, so if you close the workbench and come back it remembers what was green and red. If you open the tutorial a second time, it just resets the tutorial to its initial html.

Should there be a "Are you sure?" moment in there to ensure that it doesn't overwrite anything they've done without their knowledge? Or is that too much of an intrusion? I mean, we're just talking the state of some html here but still it could be annoying to redo some of those state changes.

bfitzpat commented:
On the plus side, doing things like writing tutorials and previewing the work is a little easier. Each Ctrl+Alt+V refreshes the window so it automatically updates from the Didact adoc/md to a HTML rendering. You just have to save the file (Ctrl+S) and hit preview again (Ctrl+Alt+V) and it reflects those changes in the same place you dragged the window to.

Not quite as good as a live preview, but close.
![previewdemo](https://user-images.githubusercontent.com/530878/106055200-4e578f00-60aa-11eb-94fe-989bca498ba7.gif)

apupier commented:
> Should there be a "Are you sure?" moment in there to ensure that it doesn't overwrite anything they've done without their knowledge? Or is that too much of an intrusion? I mean, we're just talking the state of some html here but still it could be annoying to redo some of those state changes.

I think the are you sure step could be nice if their is also a checkbox to disable it fo rnext time. I think this can be handled in another iteration.

apupier commented on src/didactManager.ts:
```diff
@@ -0,0 +1,124 @@
+		for (let index = 0; index < this._panels.length; index++) {
+			const p = this._panels[index];
+			const originalUri = p.getDidactUriPath();
+			if (testUri && originalUri) {
+				if (originalUri.toString() === testUri.toString()) {
```
the two if conditions can be merged

bfitzpat commented:
> I prefer not deactivating them without good reasons but when the rules are not followed, it is then better to deactivate them. it avoids to have noise of rule that you choose to not follow.
> It is better to introduce a type instead of disabling the rule. But you do not want. So it remains better to to disable the rule if it is not followed.
> By not creatign the type, the type-safe part of TypeScript language is not leveraged. We will hit all the troubles of non-typed languages

Though I don't disagree with you in principle, I definitely disagree as a practical matter here. Taking a look around github for other examples (even Microsoft https://github.com/search?q=deserializeWebviewPanel+user%3Amicrosoft+language%3ATypeScript+language%3ATypeScript&type=Code&ref=advsearch&l=TypeScript&l=TypeScript) nobody is making it harder than it needs to be by avoiding the any type. In our case, this typescript file is  < 25 lines of code, the purpose is clear, and overcomplicating things to avoid eslinter rules doesn't buy us anything from a practical matter since typescript boils down to javascript in the end. 

So the fact that I "do not want" to overcomplicate things is correct. If we truly want to deal with adding a type, that can be done in a follow-on code change, but at this time it seems like a lot more trouble than it's worth.

apupier reviewed (approved):
good feature. The removal of the in-house history is a very good improvement from a maintenance point of view.
The downside of the new code is the non-typed state that is used which is nto optimizing the maintainability.

apupier commented:
> In our case, this typescript file is < 25 lines of code, the purpose is clear

It is softly linked to another file where the setState is used. The day the expected inout is modified, the setStates won't be detected at compile time.
